Client/Frontend/Brave Notifications/BraveNotificationsPresenter.swift

```diff
@@ -0,0 +1,266 @@
+      guard let notification = visibleNotification, notification.view == notificationView else {
+        return
+      }
+      // Make sure to stop the dismiss timer
+      dismissTimers[notification.id]?.invalidate()
+    case .changed:
+      notificationView.transform.ty = min(0, pan.translation(in: notificationView).y)
+    case .ended:
+      let velocity = pan.velocity(in: notificationView).y
+      let y = min(panState.y, panState.y + pan.translation(in: notificationView).y)
+      let projected = project(initialVelocity: velocity, decelerationRate: UIScrollView.DecelerationRate.normal.rawValue)
+      if y + projected < 0 {
+        // must be the current visible notification
+        guard let notification = visibleNotification, notification.view == notificationView else {
+          return
+        }
+        hide(notificationView: notificationView, velocity: velocity)
+        notification.willDismiss(timedOut: false)
+        break
+      }
+      fallthrough
+    case .cancelled:
+      // must be the current visible notification
+      guard let notification = visibleNotification, notification.view == notificationView else {
+        return
+      }
+      // Re-setup timeout timer
+      if case .automatic(let interval) = notification.dismissPolicy {
+        setupTimeoutTimer(for: notification, interval: interval)
+      }
+      notificationView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+        animation.toValue = 0
+      }
+    default:
+      break
+    }
+  }
+  
+  // MARK: - Animations
+  
+  private func animateIn(adView: UIView) {
+    adView.layoutIfNeeded()
+    adView.layer.transform = CATransform3DMakeTranslation(0, -adView.bounds.size.height, 0)
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = 0
+    }
+  }
+  
+  private func animateOut(adView: UIView, velocity: CGFloat? = nil, completion: @escaping () -> Void) {
+    adView.layoutIfNeeded()
+    let y = adView.frame.minY - view.safeAreaInsets.top - adView.transform.ty
+    
+    adView.layer.springAnimate(property: kPOPLayerTranslationY, key: "translation.y") { animation, _ in
+      animation.toValue = -(view.safeAreaInsets.top + y + adView.bounds.size.height)
+      if let velocity = velocity {
+        animation.velocity = velocity
+      }
+      animation.completionBlock = { _, _ in
+        completion()
+      }
+    }
+  }
+}
+
+extension BraveNotificationsPresenter {
+  class PresenterView: UIView {
+    weak var viewController: BraveNotificationsPresenter?
+    
+    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+      guard let view = super.hitTest(point, with: event) else { return nil }
+      // Only allow tapping the ad part of this VC
+      if view.superview == viewController?.visibleNotification?.view || view == viewController?.visibleNotification?.view {
```

Comment:
Use [`isDescendant(of:)`](https://developer.apple.com/documentation/uikit/uiview/1622521-isdescendantofview) to ensure you take into account all subviews of the visible notification view

i.e. 

```swift
if let notificationView = viewController?.visibleNotification?.view, 
   view.isDecendant(of: notificationView) {
  return view
}
return nil
```